[5.6] A cache key that can be used to identify a model

paulredmond opened the pull request:
__Originally submitted in #22744, this is now based off of 5.6__ 🍸 

Based on [ActiveRecord](https://github.com/rails/rails/blob/15ef55efb591e5379486ccf53dd3e13f416564f6/activerecord/lib/active_record/integration.rb#L64), adds a `Model::cacheKey()` method that gets a stable cache key used to identify a model.

This can be used to build model cache that gets invalidated:

```php
class Article extends Model
{
    public function getCachedCommentsCountAttribute()
    {
        return Cache::rememberForever($this->cacheKey() . ':comments_count', function () {
            return $this->comments->count();
        });
    }
}
```

The `cacheKey()` method would produce the following if the model had the `updated_at` date attribute:

```
articles/1-1515693922
```

Models without an `updated_at`:

```
articles/1
```

New models:

```
articles/new
```

For more info, I [wrote about it](https://laravel-news.com/laravel-model-caching) using `Model::touch()` to invalidate cache on a model.

paulredmond committed:
A cache key that can be used to identify a model

Based on ActiveRecord, cacheKey() returns a stable cache key
that can be used to identify a model.

mfn commented on src/Illuminate/Database/Eloquent/Model.php:
```diff
@@ -1372,6 +1372,24 @@ public function getForeignKey()
+            return $this->getTable().'/new';
+        }
+
+        if (! $this->getAttribute('updated_at')) {
```
I think this isn't correct: you can configure the actual name of the updated at column (think about existing schemas with different).

I believe this needs to call `getUpdatedAtColumn().

mfn commented on src/Illuminate/Database/Eloquent/Model.php:
```diff
@@ -1372,6 +1372,24 @@ public function getForeignKey()
+            return $this->getTable().'/'.$this->getKey();
+        }
+
+        return $this->getTable().'/'.$this->getKey().'-'.$this->updated_at->timestamp;
```
Same as above

mfn commented on tests/Database/DatabaseEloquentModelTest.php:
```diff
@@ -1699,6 +1699,25 @@ public function testIsWithAnotherConnection()
+        $withDates->exists = true;
+        $this->assertEquals('stub/1-'.Carbon::now()->timestamp, $withDates->cacheKey());
+    }
+
```
A test should be added with a custom `updateAt` column for completeness.

paulredmond commented:
@mfn thank you 🙏 , I'll update it.

paulredmond committed:
Get the updated_at column and verify timestamps

Use the getUpdatedAtColumn() to get the correct column, verify that the
model uses timestamps, and that the updatedAt attribute isn't null.

Adds additional tests to ensure models with a custom updated_at work as
expected.

paulredmond committed:
Update tests

fitztrev commented:
Your article states that this technique would only hit the database when the parent is updated, but won't `cacheKey()` always do a query of its own in order to get the `updated_at`?

`cacheKey()` could never be used if someone wanted to cache the model itself. It would only work on relationships like in your example.

mfn commented:
> but won't cacheKey() always do a query of its own in order to get the updated_at?

I believe you're correct.

Something like `getAttributeFromArray()` would need to be called, which only gets it from the model attributes to not trigger that.

mfn commented:
So basically there's a problem with this approach with partially loaded models… …but I guess this is a "developers problem" to start with.

sisve commented:
A developer can copy/paste this single cacheKey() method into their own methods if they want this functionality. I see no use-case at all for this to be part of the framework. I also see future issues when people ask about cache _invalidation_ and how to use this with tags.

Don't attempt to provide caching on a framework level, it's complex enough in the application level.

taylorotwell commented:
I would indeed like to see Eloquent caching fleshed out as a package first that we could look at and evaluate.
